### PR TITLE
[WIP] Implement gradient for RGB to HSV operation.

### PR DESCRIPTION
**Pull request by kmh4321**

### CODE REVIEW

File: tensorflow/python/ops/image_grad.py [hunk 1]
```diff
@@ -153,3 +153,211 @@ def _CropAndResizeGrad(op, grad):
...
+  blues = op.inputs[0][..., 2]
+  # Output Channels
+  hue = op.outputs[0][..., 0]
+  saturation = op.outputs[0][..., 1]
+  value = op.outputs[0][..., 2]
+
+  # Mask/Indicator for max and min values of each pixel.
+  # Arbitrary assignment in case of tie breakers with R>G>B.
+  # Max values
+  red_biggest = cast((reds >= blues) & (reds >= greens), dtypes.float32)
```
Comment (feihugis) [inline tensorflow/python/ops/image_grad.py]: I'm not sure if there are more efficient ways to find the Max/Min masks.

File: tensorflow/python/ops/image_grad.py [hunk 2]
```diff
@@ -153,3 +153,211 @@ def _CropAndResizeGrad(op, grad):
...
+  green_biggest = cast((greens > reds) & (greens >= blues), dtypes.float32)
+  blue_biggest = cast((blues > reds) & (blues > greens), dtypes.float32)
+  # Min values
+  red_smallest = cast((reds < blues) & (reds < greens), dtypes.float32)
+  green_smallest = cast((greens <= reds) & (greens < blues), dtypes.float32)
+  blue_smallest = cast((blues <= reds) & (blues <= greens), dtypes.float32)
+
+  ##############################################################
+  #  Derivatives of R, G, B wrt Value slice
```
Comment (feihugis) [inline tensorflow/python/ops/image_grad.py]: Is this comment style used in other places?

File: tensorflow/python/ops/image_grad.py [hunk 3]
```diff
@@ -153,3 +153,211 @@ def _CropAndResizeGrad(op, grad):
...
+  ##############################################################
+  # Derivatives of R, G, B wrt Saturation slice
+  ##############################################################
+  # The first term in the addition is the case when the corresponding color
+  # from (r,g,b) was "MAX"
+  # -> derivative = MIN/square(MAX), MIN could be one of the other two colors
+  # The second term is the case when the corresponding color from
+  # (r,g,b) was "MIN"
+  # -> derivative = -1/MAX, MAX could be one of the other two colours.
+  ds_dr = cast(reds > 0, dtypes.float32) * add(red_biggest * \
```
Comment (feihugis) [inline tensorflow/python/ops/image_grad.py]: Could we use a helper function to compute `ds_dr`, `ds_dr_1`, ... `ds_dr_5` to avoid the repeat code?

**Review by feihugis — CHANGES_REQUESTED** [review]
This PR looks cool. Left some comments. Some tests may also need to be added.

File: tensorflow/python/ops/image_grad.py (hunk 3 above)
Comment (kmh4321) [reply tensorflow/python/ops/image_grad.py]: I thought of doing that but the formula is different each time. Each of the 5 terms has different formula for each of the three colors. 
https://en.wikipedia.org/wiki/HSL_and_HSV#From_RGB

File: tensorflow/python/ops/image_grad.py (hunk 2 above)
Comment (kmh4321) [reply tensorflow/python/ops/image_grad.py]: I don't think I have seen this, I just added to make clear sub-sections. Should I remove them?

File: tensorflow/python/ops/image_grad.py (hunk 1 above)
Comment (kmh4321) [reply tensorflow/python/ops/image_grad.py]: I was thinking of only doing one time comparision with just the max() and min() operator for each element in the tensor and then generate masks from that but the problem is that each step has to be a tensor operation and not array operation, ie it is defining a graph, not the actual operation on numbers. So this was the most convenient method. If there is a better way to do this I'm open to suggestions.

File: tensorflow/python/ops/image_grad.py (hunk 2 above)
Comment (feihugis) [reply tensorflow/python/ops/image_grad.py]: Yeah. We had better keep the comment style consistent.

File: tensorflow/python/ops/image_grad.py (hunk 3 above)
Comment (feihugis) [reply tensorflow/python/ops/image_grad.py]: Ah, I did not notice that.

File: tensorflow/python/ops/image_grad.py (hunk 1 above)
Comment (feihugis) [reply tensorflow/python/ops/image_grad.py]: Maybe you can try [tf.math.argmax](https://www.tensorflow.org/api_docs/python/tf/math/argmax).

**Pull request merged**